### PR TITLE
StreamEncoder/StreamDecoder

### PR DESCRIPTION
**Pull request by mpcsh**
Upgraded from proof-of-concept to workable API status. Tests run for files between 1 and 1000 characters long of random numbers.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -17,7 +17,7 @@ $(BUILD_STAMP): \
 	date > $@
 
 clean:
-	find lib -type f -name *.js -exec rm {} \;
+	rm -rf lib/*.js
```
Comment (maxtaco) [inline Makefile:20]: why this change?

File: src/stream.iced [hunk 2]
```diff
@@ -0,0 +1,20 @@
+stream = require('stream');
+enc = require('./encoding');
+
+exports.StreamEncoder = class StreamEncoder extends stream.Transform
+
+	constructor : (@encoder) ->
```
Comment (maxtaco) [inline src/stream.iced]: use 2 spaces per indent

File: Makefile (hunk 1 above)
Comment (mpcsh) [inline Makefile:20]: first one seemed overly complex, no real reason though

File: src/stream.iced [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+		super({highWaterMark : Math.floor(4096/encoder.in_block_len)})
```
Comment (maxtaco) [inline src/stream.iced]: can you pull out this `4096` block specification into a constant both classes share?

And actually, my formula is totally whack, and I don't see how the tests were passing with it.  It should be:

```
encoder.in_block_len * Math.ceil(desired_block_size / encoder.in_block.len)
```

Sorry I gave you the wrong formula

File: src/stream.iced (hunk 2 above)
Comment (mpcsh) [inline src/stream.iced]: done - see 6ed044c

File: src/stream.iced [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+	_transform : (chunk, encoding, callback) ->
```
Comment (maxtaco) [inline src/stream.iced]: By convention call it `cb` and not `callback`

File: src/stream.iced [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+		@push(@encoder.encode(chunk))
+		callback
```
Comment (maxtaco) [inline src/stream.iced]: This line is a noop.  You want `cb()`

Comment (maxtaco) [inline src/stream.iced]: How do tests work with this broken btw?

File: src/stream.iced [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+exports.StreamDecoder = class Streamer extends stream.Transform
+
```
Comment (maxtaco) [inline src/stream.iced:20]: class name should be `class StreamDecoder`

File: src/stream.iced [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+	constructor : (@encoder) ->
+		super({highWaterMark : Math.floor(4096/encoder.out_block_len)})
```
Comment (maxtaco) [inline src/stream.iced]: same here with the formula for highWaterMark

File: src/stream.iced [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+	_transform : (chunk, encoding, callback) ->
+		@push(@encoder.decode(chunk))
+		callback
```
Comment (maxtaco) [inline src/stream.iced]: same here with `cb()`

File: src/stream.iced [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+	_transform : (chunk, encoding, callback) ->
+		@push(@encoder.encode(chunk))
+		callback
+
+exports.StreamDecoder = class Streamer extends stream.Transform
+
+	constructor : (@encoder) ->
+		super({highWaterMark : Math.floor(4096/encoder.out_block_len)})
+
+	_transform : (chunk, encoding, callback) ->
```
Comment (maxtaco) [inline src/stream.iced]: same here with `cb()`

File: test/files/stream.iced [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+exports.test_b58_consistency = (T, cb) ->
+
+	for i in [1...1000]
+		encoder = new stream.StreamEncoder(new enc.Encoding(b58enc.alphabet, b58enc.in_block_len))
+		decoder = new stream.StreamDecoder(new enc.Encoding(b58enc.alphabet, b58enc.in_block_len))
+		encoder.pipe(decoder)
+
+		data_string = ''
+		data_string += Math.floor(Math.random()*10) for j in [0...i]
```
Comment (maxtaco) [inline test/files/stream.iced]: Get in the habit of never using `Math.random`, ever ever ever

Instead, I would do something like

``` coffee
{prng} = require 'crypto'
data = prng(i)
```

File: test/files/stream.iced [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+		data = new Buffer(data_string)
+		encoder.write(data)
+		decoded_data = decoder.read()
```
Comment (maxtaco) [inline test/files/stream.iced]: shouldn't reading from the stream block and take a callback?

File: test/files/stream.iced [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+			x += "="
+		return new Buffer x, 'base64'
+
+#==========================================================
+#These tests encode then immediately decode, and compare the result to the original text
+#=========================================================
+
+exports.test_b58_consistency = (T, cb) ->
+
+	for i in [1...1000]
```
Comment (maxtaco) [inline test/files/stream.iced]: make sure your test data goes over the block boundary, i don't think you'll get there if you do this.

File: test/files/stream.iced [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+
+exports.test_b62_consistency = (T, cb) ->
+
+	for i in [1...1000]
+		encoder = new stream.StreamEncoder(new enc.Encoding(b62enc.alphabet, b62enc.in_block_len))
+		decoder = new stream.StreamDecoder(new enc.Encoding(b62enc.alphabet, b62enc.in_block_len))
+		encoder.pipe(decoder)
+
+		data_string = ''
+		data_string += Math.floor(Math.random()*10) for j in [0...i]
```
Comment (maxtaco) [inline test/files/stream.iced]: See above for generating this.

File: test/files/stream.iced (hunk 12 above)
Comment (maxtaco) [inline test/files/stream.iced]: `for i in [1...5000]` ?

File: test/files/stream.iced [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+	for i in [1...1000]
+		encoder = new stream.StreamEncoder(new enc.Encoding(b58enc.alphabet, b58enc.in_block_len))
+		decoder = new stream.StreamDecoder(new enc.Encoding(b58enc.alphabet, b58enc.in_block_len))
+		encoder.pipe(decoder)
+
+		data_string = ''
+		data_string += Math.floor(Math.random()*10) for j in [0...i]
+
+		data = new Buffer(data_string)
+		encoder.write(data)
```
Comment (maxtaco) [inline test/files/stream.iced]: Don't write all of the data at once, i would do so in a loop, to simulate the way it's actually going to be used.

Comment (maxtaco) [inline test/files/stream.iced]: Maybe?

``` coffee
while data.length
   j = prng(0)[1]
   encoder.write data[0...j]
   data = data[j...]
   await setTimeout defer(), 0
```

File: test/files/stream.iced [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+    return new Buffer x, 'base64'
+
+#==========================================================
+#These tests encode then immediately decode, and compare the result to the original text
+#=========================================================
+
+exports.test_b58_consistency = (T, cb) ->
+
+  for i in [1...1000]
+    encoder = new stream.StreamEncoder(new enc.Encoding(b58enc.alphabet, b58enc.in_block_len))
```
Comment (maxtaco) [inline test/files/stream.iced]: why doesn't it just work to do?

`encoder = new stream.StreamEncoder(b58enc)`?

File: src/stream.iced [hunk 16]
```diff
@@ -1,20 +1,22 @@
...
+stream = require('stream')
+enc = require('./encoding')
+
+desired_block_size = 4096
 
 exports.StreamEncoder = class StreamEncoder extends stream.Transform
 
   constructor : (@encoder) ->
-    super({highWaterMark : Math.floor(4096/encoder.in_block_len)})
+    super({highWaterMark : encoder.in_block_len*Math.ceil(desired_block_size/encoder.in_block_len)})
```
Comment (maxtaco) [inline src/stream.iced]: can you factor out this arithmetic since it's used twice?

File: test/files/stream.iced [hunk 17]
```diff
@@ -17,104 +18,89 @@ b64stripped =
...
+    data = prng(i)
     encoder.write(data)
     decoded_data = decoder.read()
 
     T.equal(data, decoded_data, "inconsistency found: data=#{data}")
-  cb()
+  callback()
 
-exports.test_b64_consistency = (T, cb) ->
+exports.test_b64_consistency = (T, callback) ->
```
Comment (maxtaco) [inline test/files/stream.iced]: Use `cb` everywhere, not `callback`, since our style is the former.

File: test/files/stream.iced (hunk 14 above)
Comment (mpcsh) [inline test/files/stream.iced]: Working on this right now (see adb844979890bfc2ce94d87875cf4b19f6c246a7)

**Comment by mpcsh** [comment]
All comments from Max have now been addressed, tests are passing and performance is much improved.

File: test/files/stream.iced [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+
+#==========================================================
+#Helper functions
+#==========================================================
+
+loop_limit = 5000
+loop_skip = 27
+
+stream_random_data = (strm, len) ->
+	data = prng(len)
```
Comment (maxtaco) [inline test/files/stream.iced]: using tabs rather than spaces?

File: src/stream.iced [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+  input_length*Math.floor(desired_block_size/input_length)
+
+exports.StreamEncoder = class StreamEncoder extends stream.Transform
+
+  constructor : (@encoder) ->
+    super({highWaterMark : calculate_block_size(encoder.in_block_len)})
+
+  _transform : (chunk, encoding, cb) ->
+    @push(@encoder.encode(chunk))
+    @read(0)
```
Comment (maxtaco) [inline src/stream.iced]: what's this here for?

File: src/stream.iced [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+    cb()
+
+exports.StreamDecoder = class StreamDecoder extends stream.Transform
+
+  constructor : (@encoder) ->
```
Comment (maxtaco) [inline src/stream.iced]: `s/encoder/decoder/`

File: test/files/stream.iced [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+test_bx_consistency = (T, base, len) ->
+	encoding = encoding_for_base(base)
+	encoder = new stream.StreamEncoder(encoding)
+	decoder = new stream.StreamDecoder(encoding)
+
+	encoder.pipe(decoder)
+
+	data = stream_random_data(encoder, len)
+	decoded_data = decoder.read()
+	leftovers = encoder.read(0)
```
Comment (maxtaco) [inline test/files/stream.iced]: what do you expect to come out here?

Comment (maxtaco) [inline test/files/stream.iced]: and/or do you want to put some assertions on it?

**Comment by maxtaco** [comment]
looking a lot better, a few more questions...

File: test/files/stream.iced (hunk 18 above)
Comment (mpcsh) [inline test/files/stream.iced]: Whoops! Fixed in 699240e6146d65c6dc7813189408f32605cc3837

File: src/stream.iced (hunk 19 above)
Comment (mpcsh) [inline src/stream.iced]: I'm assuming you're talking about the `@read(0)` line - it refreshes the internal read buffer. This is what fixed the truncation bug where we weren't getting everything back after encode->decode operations that went past the highWaterMark. From the node API: https://nodejs.org/api/stream.html#stream_readable_read_0

File: src/stream.iced (hunk 20 above)
Comment (mpcsh) [inline src/stream.iced]: fixed in b2f26c26d34cff3aa699c3e4160ec2d0c31e8e13

File: test/files/stream.iced (hunk 21 above)
Comment (mpcsh) [inline test/files/stream.iced]: so the consistency test takes a bunch of random data (returned by the stream_random_data function as a buffer for easy comparison), encodes it, then decodes it, and compares the encoded->decoded buffer against the original. The assertion would be that the two are identical, which is what line 33 handles.

File: src/stream.iced (hunk 19 above)
Comment (maxtaco) [inline src/stream.iced]: Those docs say apps should almost never need to do this, but we're doing it on every read.

Comment (mpcsh) [inline src/stream.iced]: Yes, but they don't consider it bad practice - it didn't hurt performance either. They key is this:
"If the internal read buffer is below the highWaterMark, and the stream is not currently reading, then calling stream.read(0) will trigger a low-level stream._read() call."
I think the reason for their "never need to do this" remark is that in most cases, it would be simpler to either wait for more data to come onto the stream, change the highWaterMark parameter, or write zeros until the highWaterMark is reached. We can't do any of those. I also think they were wrong: more than a few implementations of TransformStreams called `read(0)` in their `_transform` functions

Comment (maxtaco) [inline src/stream.iced]: how big are the buffers you're getting back in your tests?

Comment (mpcsh) [inline src/stream.iced]: The whole stream is placed into a buffer (length depends on the current iteration) for easy comparison with T.equals - running T.equals on individual chunks is possible, but I wasn't sure it was necessary

Comment (maxtaco) [inline src/stream.iced]: just add console.logs on your incoming chunk lenghts.

**Comment by mpcsh** [comment]
Okay, so turns out the highWaterMark issue wasn't an issue with the StreamEncoder/Decoder implementation, but rather with how I was consuming the stream in my tests. I created a StreamToBuffer class that properly handles backpressure and just dumps all data into an internal buffer object which can then be accessed. All corner cases are passing now:
- [x] Partial chunks (either a message smaller than the block size or leftovers from a message that didn't evenly divide into the block size)
- [x] Larger-than-highWaterMark data: the streams can now handle arbitrarily large amounts of incoming data
- [x] Output is verified to be correct: checked against the stock base64 function, as well as checking whether or not encoding by random intervals is the same as encoding the whole thing at once

File: src/stream.iced [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+desired_high_water_mark = 4096
+calculate_high_water_mark = (input_length) ->
+  input_length*Math.floor(desired_high_water_mark/input_length)
+
+exports.StreamEncoder = class StreamEncoder extends stream.Transform
+
+  constructor : (@encoder) ->
+    @extra = null
+    @block_size = @encoder.in_block_len
+    super({highWaterMark : calculate_high_water_mark(@block_size)})
```
Comment (maxtaco) [inline src/stream.iced]: Do you even need to specify `highWaterMark` anymore?  It doesn't seem like it was buying us much since it wasn't doing what we thought it was doing anyways.

File: src/stream.iced [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+    if remainder isnt 0
+      @extra = chunk.slice(chunk.length-remainder)
+      chunk = chunk.slice(0, chunk.length-remainder)
+
+    @push(@encoder.encode(chunk))
+    cb()
+
+  _flush : (cb) ->
+    if @extra then @push(@encoder.encode(@extra))
+    @push(new Buffer(''))
```
Comment (maxtaco) [inline src/stream.iced]: why is this needed? Also, I think it's slightly preferable to do `(new Buffer [])`

File: test/files/stream.iced [hunk 24]
```diff
@@ -0,0 +1,139 @@
...
+
+
+#=========================================================
+#These tests check whether or not the encoding is valid
+#=========================================================
+
+exports.test_b64_output = (T, cb) ->
+  for i in [1...loop_limit] by loop_skip
+    test_bx_output(T, 64, i, b64stock.encode)
+  cb()
```
Comment (maxtaco) [inline test/files/stream.iced:115]: can you also test b62 test output via `test_bx_output`?

File: src/stream.iced [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+    if remainder isnt 0
+      @extra = chunk.slice(chunk.length-remainder)
+      chunk = chunk.slice(0, chunk.length-remainder)
+
+    @push(@decoder.decode(chunk))
+    cb()
+
+  _flush : (cb) ->
+    if @extra then @push(@decoder.decode(@extra))
+    @push(new Buffer(''))
```
Comment (maxtaco) [inline src/stream.iced]: same as above for style preference with empty buffer.

File: src/stream_to_buffer.iced [hunk 26]
```diff
@@ -0,0 +1,14 @@
+stream = require('stream')
+
+exports.StreamToBuffer = class StreamToBuffer extends stream.Transform
+
+  constructor : (options) ->
+    @buf = new Buffer('')
```
Comment (maxtaco) [inline src/stream_to_buffer.iced]: same as above for style preference with empty buffer.

File: src/stream.iced [hunk 27]
```diff
@@ -0,0 +1,56 @@
+stream = require('stream')
+enc = require('./encoding')
+
+desired_high_water_mark = 4096
```
Comment (maxtaco) [inline src/stream.iced]: we can probably get rid of this stuff too?

File: src/stream.iced [hunk 28]
```diff
@@ -0,0 +1,56 @@
...
+calculate_high_water_mark = (input_length) ->
+  input_length*Math.floor(desired_high_water_mark/input_length)
```
Comment (maxtaco) [inline src/stream.iced]: and nuke this too?

File: src/stream.iced (hunk 22 above)
Comment (mpcsh) [inline src/stream.iced]: It'll work with whatever `highWaterMark` you supply. I left it in because the default is 16k, which seems like a bit much.

File: src/stream.iced (hunk 23 above)
Comment (mpcsh) [inline src/stream.iced]: Oh sorry, this was just random tinkering and it turned out to not do anything. Forgot to remove it.

File: src/stream.iced (hunk 22 above)
Comment (maxtaco) [inline src/stream.iced]: ok, cool, let's stay with the defaults then, 16k is fine

File: src/stream.iced (hunk 28 above)
Comment (mpcsh) [inline src/stream.iced]: Just tested with defaults and it doesn't make an appreciable difference, removed in 766fe43f19493f589d9c0def12c5187062115c2c

File: test/files/stream.iced (hunk 24 above)
Comment (mpcsh) [inline test/files/stream.iced:115]: d6f8595

**Comment by maxtaco** [comment]
OK, great work, 🚢 it!
